crates/ibc/src/core/context.rs

```diff
@@ -1114,4 +1169,121 @@ mod val_exec_ctx {
 
         Ok(())
     }
+
+    fn recv_packet_validate<ValCtx>(ctx_b: &ValCtx, msg: MsgRecvPacket) -> Result<(), ContextError>
+    where
+        ValCtx: ValidationContext,
+    {
+        // Note: this contains the validation for `write_acknowledgement` as well.
+        recv_packet::validate(ctx_b, &msg)
+
+        // nothing to validate with the module, since `onRecvPacket` cannot fail.
+    }
+
+    fn recv_packet_execute<ExecCtx>(
+        ctx_b: &mut ExecCtx,
+        module_id: ModuleId,
+        msg: MsgRecvPacket,
+    ) -> Result<(), ContextError>
+    where
+        ExecCtx: ExecutionContext,
+    {
+        let chan_port_id_on_b = (msg.packet.port_on_b.clone(), msg.packet.chan_on_b.clone());
+        let chan_end_on_b = ctx_b.channel_end(&chan_port_id_on_b)?;
+
+        // Check if another relayer already relayed the packet.
+        // We don't want to fail the transaction in this case.
+        {
+            let packet_already_received = match chan_end_on_b.ordering {
+                // Note: ibc-go doesn't make the check for `Order::None` channels
+                Order::None => false,
+                Order::Unordered => {
+                    let packet = msg.packet.clone();
+
+                    ctx_b
+                        .get_packet_receipt(&(packet.port_on_b, packet.chan_on_b, packet.sequence))
+                        .is_ok()
+                }
+                Order::Ordered => {
+                    let next_seq_recv = ctx_b.get_next_sequence_recv(&chan_port_id_on_b)?;
+
+                    // the sequence number has already been incremented, so
+                    // another relayer already relayed the packet
+                    msg.packet.sequence < next_seq_recv
+                }
+            };
+
+            if packet_already_received {
+                return Ok(());
+            }
+        }
```

Comment:
It's not; notice that what we're trying to achieve here seeing if we should return early or not (or, in other words, don't do any writes). Does that make sense?

Reply:
> A duplicate msg is not valid to process.

Actually it is. With every packet to relay, relayers are in a race to relay the packet, and they'll all try to send the same `recvPacket` transaction at the same time. But only the first one will be accepted by the chain. Previously, ibc-go didn't have that "already received" check, and so all the relayers' transactions (except the first successful one) would error, which meant that other messages in the transaction would be discarded, and increased fees for the relayer (I think). But there's no point in penalizing relayers in this way, so ibc-go made this case a valid `recvPacket` message. See this ibc-go issue for more context: https://github.com/cosmos/ibc-go/issues/245.

> multiple relayers try to manipulate and attack by sending tons of messages

How is this an attack? What exactly would be manipulated? AFAIK, they're not gaining anything, because they spent money on the transaction, but the transaction ended up being a no-op and not changing any of the blockchain state.

Reply:
>  Previously, ibc-go didn't have that "already received" check, and so all the relayers' transactions (except the first successful one) would error, 

I don't see it was considered as a valid msg. Instead of returning errors they return no-ops. 
Let's look at it from another side: which of the execution steps (state change, store, emit, log) are done on a redundant message? All of them are dropped as they are "not valid for execution"

(I am discussing this to clarify the validation and execution boundaries. All in all, It's not a big deal)

> How is this an attack? What exactly would be manipulated? AFAIK, they're not gaining anything, because they spent money on the transaction, but the transaction ended up being a no-op and not changing any of the blockchain state.

Something like DDoS attack, Getting the network down for a while by consuming the chain or relayers processing capacity to incur a hack by manipulating some other parts at that time.

Reply:
> Let's look at it from another side: which of the execution steps (state change, store, emit, log) are done on a redundant message? All of them are dropped as they are "not valid for execution"

That's right; none are executed. Stated differently, sending a `recvPacket` msg for a packet that has already been received is a valid transaction which doesn't change the state. The reason for it to be a valid message is simply so that relayers don't have to pay extra fees for doing exactly what they're supposed to do. If we failed the validation, then they would pay extra fees; so instead we succeed the transaction, but don't change the state machine.

> Something like DDoS attack, Getting the network down for a while by consuming the chain or relayers processing capacity to incur a hack by manipulating some other parts at that time.

Can you expand on how that attack would work? How is this any different than someone spamming the chain with any other valid transaction (e.g. sending tokens back and forth between their own accounts)?

Reply:
Per our discussion and giving it more thoughts, it makes sense :pray: